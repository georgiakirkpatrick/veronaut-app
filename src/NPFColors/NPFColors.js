import React from 'react'
import '@fortawesome/react-fontawesome'
import FormPage from '../FormPage/FormPage'
import FormButton from '../FormButton/FormButton'
import FormDropdown from '../FormDropdown/FormDropdown'
import FormFieldset from '../FormFieldset/FormFieldset'
import FormPromptWithSub from '../FormPromptWithSub/FormPromptWithSub'
import NPFFooter from '../NPFFooter/NPFFooter';
import FormTextInput from '../FormTextInput/FormTextInput'
import formData from '../FORM_DATA'

const NPFColors = props => {

    const {
        colorFieldsets,
        currentPage,
        setColorFieldsets,
        setPage,
    } = props

    const addColor = () => {
        setColorFieldsets(
            [
                ...colorFieldsets, 
                {
                    name: '', 
                    descriptionId: 0, 
                    swatchUrl: '',
                    imageUrls: ['']
                }
            ]
        )
    }

    const colorsChangeInput = (index, event) => {
        const values = [...colorFieldsets]
        console.log('values', values)
        values[index][event.target.name] = event.target.value
        console.log('values[index][event.target.name]', values[index][event.target.name])

        setColorFieldsets(values)
        console.log('values', values)

    }

    const makeColorOptions = () => {
        const defaultOption = {
            id: 0,
            text: 'Select a color',
            value: 0
        }
        
        const colors = formData.colors.map(color => (
            {
                id: color.id,
                text: color.text,
                value: color.id
            }
        ))

        return [
            defaultOption,
            ...colors
        ]
    }

    const nextButton = () => {
        const missingFields = []

        const requiredFields = [
            'descriptionId'
        ]

        colorFieldsets.length > 1 && requiredFields.push('name')

        colorFieldsets.forEach(fieldset => {
            requiredFields.forEach(field => {
                if (fieldset[field] === 0 || fieldset[field] === '') {
                    missingFields.push(field)
                }
            })
        })

        if (missingFields.length >= 1) {
            alert(`Please enter a description for each color option.  Remove any unneeded color fields by clicking 'remove'.`)
        } else if (missingFields.length === 0) {
            setPage(currentPage + 1)
        }
    }
    
    // const nextButton = () => {setPage(currentPage + 1)}
    
    const removeColor = (index) => {
        const values = [...colorFieldsets]
        values.splice(index, 1)
        setColorFieldsets(values)
    }

    return (
        <div id='colors'>
            <FormPage title='Colors'>
                <div id='color-inputs'>
                    <FormPromptWithSub 
                        prompt='Enter all color options listed on the product webpage'
                        promptSubtitle='If the product has multiple colors, what is the dominant color?'
                    />
                    {colorFieldsets.map((colorFieldset, index) => (
                        <FormFieldset key={index} className='NewProductForm__fieldset'>
                            <button 
                                className='NewProductForm__remove' 
                                type='button' 
                                onClick={() => removeColor(index)}
                            >
                                REMOVE
                            </button>
                            <FormTextInput 
                                id={'color-name' + index}
                                name='name'
                                prompt='Color name'
                                currentValue={colorFieldset.name} 
                                handleChange={event => {
                                    colorsChangeInput(index, event)
                                }}
                            />
                            <FormDropdown 
                                id={'color-description' + index} 
                                name='descriptionId'
                                prompt='Color description'
                                // 'If the product has multiple colors, what is the dominant color?'
                                options={makeColorOptions()}
                                currentValue={colorFieldset.descriptionId} 
                                handleChange={event => {
                                    colorsChangeInput(index, event)
                                }}
                            />
                        </ FormFieldset>
                    ))}
                </div>
    
                <FormButton 
                    buttonText='ADD A COLOR' 
                    handleClick={() => addColor() } 
                />                
            </FormPage>
            <NPFFooter
                buttons='prevNext'
                previousButton={() => setPage(currentPage - 1)} 
                nextButton={() => {
                    nextButton()
                }}
            />
        </div>
    )
}

NPFColors.defaultProps = {
    currentPage: 0,
    colorFieldsets: [],
    setColorFieldsets: () => {},
    setPage: () => {},
}

export default NPFColors