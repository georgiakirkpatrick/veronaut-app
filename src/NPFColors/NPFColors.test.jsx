import React from 'react'
import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import NPFColors from './NPFColors'

describe('NPFColors', () => {
  it ('renders without crashing', () => {
    render(<NPFColors />)
    screen.debug()
  })

  it ('render the UI as expected', () => {
    const tree = render(<NPFColors />)
    expect(tree).toMatchSnapshot()
  })
})