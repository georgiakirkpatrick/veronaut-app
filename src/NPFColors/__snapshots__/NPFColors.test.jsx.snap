// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NPFColors > render the UI as expected 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="colors"
      >
        <section
          class="FormPage"
        >
          <h1>
            Colors
          </h1>
          <form
            autocomplete="off"
            id="form-page"
          >
            <div
              class="FormPromptWithSub"
            >
              <p
                class="FormPromptWithSub__prompt"
              >
                Enter all color options listed on the product webpage
              </p>
              <p
                class="FormPromptWithSub__subtitle"
              >
                If a product color option contains multiple colors, what is the dominant color?
              </p>
            </div>
            <button
              class="FormButton"
              type="button"
            >
              Add a color option
            </button>
          </form>
          <div />
        </section>
        <footer
          class="NPFFooter"
        >
          <button
            class="NPFFooter__previous"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-left NPFFooter__left-arrow"
              data-icon="angle-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
                fill="currentColor"
              />
            </svg>
            <span>
              PREVIOUS
            </span>
          </button>
          <button
            class="NPFFooter__next"
            type="button"
          >
            <span>
              NEXT
            </span>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-angle-right NPFFooter__right-arrow"
              data-icon="angle-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
                fill="currentColor"
              />
            </svg>
          </button>
        </footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="colors"
    >
      <section
        class="FormPage"
      >
        <h1>
          Colors
        </h1>
        <form
          autocomplete="off"
          id="form-page"
        >
          <div
            class="FormPromptWithSub"
          >
            <p
              class="FormPromptWithSub__prompt"
            >
              Enter all color options listed on the product webpage
            </p>
            <p
              class="FormPromptWithSub__subtitle"
            >
              If a product color option contains multiple colors, what is the dominant color?
            </p>
          </div>
          <button
            class="FormButton"
            type="button"
          >
            Add a color option
          </button>
        </form>
        <div />
      </section>
      <footer
        class="NPFFooter"
      >
        <button
          class="NPFFooter__previous"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-left NPFFooter__left-arrow"
            data-icon="angle-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
              fill="currentColor"
            />
          </svg>
          <span>
            PREVIOUS
          </span>
        </button>
        <button
          class="NPFFooter__next"
          type="button"
        >
          <span>
            NEXT
          </span>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-right NPFFooter__right-arrow"
            data-icon="angle-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"
              fill="currentColor"
            />
          </svg>
        </button>
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
